src/lib.rs

```diff
@@ -19,68 +26,48 @@ const USECS_PER_MS: i64 = 1_000;
 
 const STALE_NAN: u64 = 0x7ff0000000000002;
 
-// prom divides time into sliding windows of fixed size, e.g.
-// |  5 seconds  |  5 seconds  |  5 seconds  |  5 seconds  |  5 seconds  |
-// we take the first and last values in that bucket and uses `last-first` as the
-// value for that bucket.
-//  | a b c d e | f g h i | j   k |   m    |
-//  |   e - a   |  i - f  | k - j | <null> |
 #[pg_extern(immutable, parallel_safe)]
-pub fn prom_delta_transition(
-    state: Option<Internal<GapfillDeltaTransition>>,
+pub fn prom_rate_transition(
+    state: Internal,
     lowest_time: TimestampTz,
     greatest_time: TimestampTz,
-    step_size: Milliseconds, // `prev_now - step` is where the next window starts
-    range: Milliseconds,     // the size of a window to delta over
+    step_size: Milliseconds,
+    range: Milliseconds, // the size of a window to calculate over
     time: TimestampTz,
     val: f64,
     fc: pg_sys::FunctionCallInfo,
-) -> Option<Internal<GapfillDeltaTransition>> {
-    unsafe {
-        in_aggregate_context(fc, || {
-            if time < lowest_time || time > greatest_time {
-                error!("input time less than lowest time")
-            }
-
-            let mut state = state.unwrap_or_else(|| {
-                let state: Internal<_> = GapfillDeltaTransition::new(
-                    lowest_time,
-                    greatest_time,
-                    range,
-                    step_size,
-                    false,
-                    false,
-                )
-                .into();
-                state
-            });
-
-            state.add_data_point(time, val);
-
-            Some(state)
-        })
-    }
+) -> Internal {
+    prom_rate_transition_inner(
+        unsafe { state.to_inner() },
+        lowest_time.into(),
+        greatest_time.into(),
+        step_size,
+        range,
+        time.into(),
+        val,
+        fc,
+    )
+    .internal()
 }
 
-#[pg_extern(immutable, parallel_safe)]
-pub fn prom_rate_transition(
-    state: Option<Internal<GapfillDeltaTransition>>,
-    lowest_time: TimestampTz,
-    greatest_time: TimestampTz,
+pub fn prom_rate_transition_inner(
```

Comment:
do we really need this split between the inner and main function all over the place? Was this just for schema generation?

Reply:
I think that schema generation is part of the motivation, yes. As far as I understand, we need the `Internal` type in the API of the method (as an arg and return type), for correct schema generation. Additionally, we have the generic `Inner<T>` wrapper which we use internally. In the main function we translate `Internal` to `Inner<T>`.

I assume that @JLockerman can give a better answer here.